From ef0bddda0a46b6560db3df684931358c8f964e64 Mon Sep 17 00:00:00 2001
From: Giovanni Mascellani <gmascellani@codeweavers.com>
Date: Mon, 7 Dec 2020 09:49:53 +0100
Subject: [PATCH 0306/1562] winex11.drv: Dump keysyms and translations for all
 keys.

Dump all we can see about the user keyboard, so that their +keyboard
logs can be used to fix layout tables.

Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=30984
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=45605
CW-Bug-Id: #16793
---
 dlls/winex11.drv/keyboard.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/dlls/winex11.drv/keyboard.c b/dlls/winex11.drv/keyboard.c
index b561bd997b6..a0e8047605c 100644
--- a/dlls/winex11.drv/keyboard.c
+++ b/dlls/winex11.drv/keyboard.c
@@ -1500,6 +1500,19 @@ X11DRV_KEYBOARD_DetectLayout( Display *display )
       for (i = 0; i < syms; i++) {
         if (!(keysym = XkbKeycodeToKeysym( display, keyc, 0, i ))) continue;
         ckey[keyc][i] = keysym_to_char(keysym);
+        if (TRACE_ON(keyboard))
+        {
+            char buf[32];
+            WCHAR bufW[32];
+            int len, lenW;
+            KeySym orig_keysym = keysym;
+            len = XkbTranslateKeySym(display, &keysym, 0, buf, sizeof(buf), NULL);
+            lenW = ntdll_umbstowcs(buf, len, bufW, ARRAY_SIZE(bufW));
+            if (lenW < ARRAY_SIZE(bufW))
+                bufW[lenW] = 0;
+            TRACE("keycode %u, index %d, orig_keysym 0x%04lx, keysym 0x%04lx, buf %s, bufW %s\n",
+                    keyc, i, orig_keysym, keysym, debugstr_a(buf), debugstr_w(bufW));
+        }
       }
   }
 
-- 
2.45.2

